File: BAEKJOON/2098 - 외판원 순회/외판원 순회.cpp
#include <iostream>
#include <vector>
#include <algorithm>

const int INF = 1e9;  // 무한대를 표현하기 위한 큰 수
int N;
std::vector<std::vector<int>> W;
std::vector<std::vector<int>> dp;

int tsp(int current, int visited) {
    if (visited == (1 << N) - 1) {  // 모든 도시를 방문한 경우
        return W[current][0] == 0 ? INF : W[current][0];  // 시작 도시로 돌아갈 수 있는지 확인
    }
    
    int &ret = dp[current][visited];
    if (ret != -1) return ret;  // 이미 계산된 값이 있으면 반환
    
    ret = INF;
    for (int i = 0; i < N; ++i) {
        if (visited & (1 << i)) continue;  // 이미 방문한 도시는 패스
        if (W[current][i] == 0) continue;  // 갈 수 없는 도시는 패스
        ret = std::min(ret, W[current][i] + tsp(i, visited | (1 << i)));  // 최솟값 갱신
    }
    
    return ret;
}

int main() {
    std::cin >> N;
    W = std::vector<std::vector<int>>(N, std::vector<int>(N));
    dp = std::vector<std::vector<int>>(N, std::vector<int>(1 << N, -1));
    
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            std::cin >> W[i][j];
        }
    }
    
    int result = tsp(0, 1);
    if (result >= INF) result = -1;  // 결과가 무한대이면 경로가 없음을 의미
    
    std::cout << result << std::endl;
    return 0;
}

File: BAEKJOON/2098 - 외판원 순회/외판원 순회.py
import sys

def tsp(n, W):
    # 상수 정의
    INF = sys.maxsize  # 무한대를 나타내기 위해 최대 정수를 사용
    
    # DP 테이블 초기화
    dp = [[INF] * (1 << n) for _ in range(n)]  # dp[i][mask]는 i번째 도시에 있고, mask 상태일 때의 최소 비용
    
    # 출발지에서 출발하는 비용은 0
    dp[0][1] = 0  # 첫 번째 도시에서 출발하는 비용은 0 (출발지는 항상 0번 도시)
    
    # 모든 도시 순회
    for mask in range(1 << n):  # 모든 가능한 방문 상태를 나타내는 비트마스크
        for i in range(n):  # 현재 도시
            if mask & (1 << i):  # 현재 비트마스크에서 i번째 도시를 방문했는지 확인
                for j in range(n):  # 다음 방문할 도시
                    if not (mask & (1 << j)) and W[i][j] != 0:  # j번째 도시를 방문하지 않았고, i에서 j로 가는 경로가 존재할 때
                        new_mask = mask | (1 << j)  # j번째 도시를 방문하는 새로운 비트마스크
                        # 새로운 상태로 갱신
                        dp[j][new_mask] = min(dp[j][new_mask], dp[i][mask] + W[i][j])  # i에서 j로 가는 비용을 추가하여 최소 비용 갱신
    
    # 모든 도시를 방문하고 출발지로 돌아오는 최소 비용 찾기
    min_cost = INF  # 최소 비용 초기화
    for i in range(1, n):  # 출발지(0번 도시)를 제외한 모든 도시에서 출발
        if W[i][0] != 0:  # i에서 출발지로 돌아오는 경로가 존재할 때
            min_cost = min(min_cost, dp[i][(1 << n) - 1] + W[i][0])  # 모든 도시를 방문하고 출발지로 돌아오는 최소 비용 갱신
    
    return min_cost  # 최종 최소 비용 반환

# 입력 받기
n = int(input())  # 도시의 수 입력
W = [list(map(int, input().split())) for _ in range(n)]  # 비용 행렬 입력

# TSP 문제 해결
result = tsp(n, W)  # 최소 비용 계산

# 결과 출력
print(result)  # 최소 비용 출력
